src/framebuffer.rs

```diff
@@ -0,0 +1,707 @@
+                let x = p.x as usize;
+                let y = p.y as usize;
+
+                self.data[y * WIDTH + x] = c.into().into_inner();
+            }
+        }
+    }
+}
+
+impl<C, BO, const WIDTH: usize, const HEIGHT: usize, const N: usize> DrawTarget
+    for Framebuffer<C, RawU8, BO, WIDTH, HEIGHT, N>
+where
+    C: PixelColor<Raw = RawU8> + Into<RawU8>,
+{
+    type Color = C;
+    type Error = Infallible;
+
+    fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+    where
+        I: IntoIterator<Item = Pixel<Self::Color>>,
+    {
+        for Pixel(p, c) in pixels {
+            self.set_pixel(p, c);
+        }
+
+        Ok(())
+    }
+}
+
+macro_rules! impl_bytes {
+    ($raw_type:ty, $bo_type:ty, $to_bytes_fn:ident) => {
+        impl<C, const WIDTH: usize, const HEIGHT: usize, const N: usize>
+            Framebuffer<C, $raw_type, $bo_type, WIDTH, HEIGHT, N>
+        where
+            C: PixelColor + Into<$raw_type>,
+        {
+            /// Sets the color of a pixel.
+            ///
+            /// Trying to set a pixel outside the framebuffer is a noop.
+            pub fn set_pixel(&mut self, p: Point, c: C) {
+                const BYTES_PER_PIXEL: usize = <$raw_type>::BITS_PER_PIXEL / 8;
+
+                if let (Ok(x), Ok(y)) = (usize::try_from(p.x), usize::try_from(p.y)) {
+                    if x < WIDTH && y < HEIGHT {
+                        let x = p.x as usize;
+                        let y = p.y as usize;
+
+                        let index = (y * WIDTH + x) * BYTES_PER_PIXEL;
+
+                        self.data[index..index + BYTES_PER_PIXEL]
+                            .copy_from_slice(&c.into().$to_bytes_fn());
+                    }
+                }
+            }
+        }
+
+        impl<C, const WIDTH: usize, const HEIGHT: usize, const N: usize> DrawTarget
+            for Framebuffer<C, $raw_type, $bo_type, WIDTH, HEIGHT, N>
+        where
+            C: PixelColor<Raw = $raw_type> + Into<$raw_type>,
+        {
+            type Color = C;
+            type Error = Infallible;
+
+            fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+            where
+                I: IntoIterator<Item = Pixel<Self::Color>>,
+            {
+                for Pixel(p, c) in pixels {
+                    self.set_pixel(p, c);
+                }
+
+                Ok(())
+            }
+        }
+    };
+
+    ($raw_type:ty) => {
+        impl_bytes!($raw_type, LittleEndian, to_le_bytes);
+        impl_bytes!($raw_type, BigEndian, to_be_bytes);
+    };
+}
+
+impl_bytes!(RawU16);
+impl_bytes!(RawU24);
+impl_bytes!(RawU32);
+
+impl<C, R, BO, const WIDTH: usize, const HEIGHT: usize, const N: usize> OriginDimensions
+    for Framebuffer<C, R, BO, WIDTH, HEIGHT, N>
+{
+    fn size(&self) -> Size {
+        Size::new(WIDTH as u32, HEIGHT as u32)
+    }
+}
+
+/// Calculate the framebuffer generic constants.
+#[macro_export]
+macro_rules! framebuffer {
```

Comment:
Do we need these macros beyond some convenience for the user? This is another API item we'd end up removing in the future once more const features are stablised. I think I'd prefer to export `buffer_size` and rely on our docs to help users write the correct type definitions.

I'm also enjoying not having any exported macros in e-g 😁

Reply:
> Do we need these macros beyond some convenience for the user?

No.

> I'm also enjoying not having any exported macros in e-g grin

OK. I've made the macro private inside the tests module.